### ISSUE TITLE
oclvanitygen not properly saturating all GPUs.

### ISSUE THREAD

**Issue by SharkWipf** [exploitagency/vanitygen-plus#103]
While running on a single GPU properly saturates the device, running on a larger amount of GPUs leaves many GPUs undersaturated (some even stay at 0% saturation in nvidia-smi).
Command used (Pull request for STAK will follow soon): 
`./oclvanitygen -C STAK -k -D 0:0 -D 0:1 -D 0:2 -D 0:3 -D 0:4 -D 0:5 -D 0:6 -D 0:7 -o results.txt -i SharkWipf` 

Hardware: Dual Xeon E5-2620 v4, 8x GTX 1080 Ti, dual root complex.

Running oclvanitygen for each GPU separately properly saturates all GPUs (but risks running the same hashes multiple times of course). CPU is not the bottleneck.

**Comment by ilovelinux** [comment]
How much is the hashrate with one GPU and with multiple gpu?

**Comment by SharkWipf** [comment]
1 GPU: ~102 Mkey/s;
2 GPUs: ~185 Mkey/s;
3 GPUs: ~260 Mkey/s;
4 GPUs: ~335 Mkey/s;
5 GPUs: ~365 Mkey/s;
6 GPUs: ~395 Mkey/s;
7 GPUs: ~450 Mkey/s;
All 8 GPUs: ~460 Mkey/s.

1 GPU has a fairly stable hashrate but the more GPUs are added the heavier the fluctuations in hashrate get. With 8 GPUs the hashrate seems to fluctuate wildly between ~420 Mkey/s and ~500 Mkey/s.

**Comment by SharkWipf** [comment]
#27 might be related, after about an hour Vanitygen-plus stops working entirely, shows 100% saturation on 1 GPU, nothing on other GPUs and hashrates stop updating.

**Comment by chriswhocodes** [comment]
I reported the same issue in the project this is forked from: https://github.com/samr7/vanitygen/issues/96 after a few hours with multiple GPUs it will saturate 1 GPU at 100% and nothing on the others. No further progress reported to standard out. Feels like a race condition in the code that distributes work to the GPUs. Still investigating by adding debug.

**Comment by adminnu** [comment]
maybe your comment applies to my problem, is it so?
https://github.com/exploitagency/vanitygen-plus/issues/141

**Comment by k06a** [comment]
@SharkWipf did you think of bruteforcing Bitcoin addresses for money with `oclvanitygen`? How many Mkey/s you have in total?

**Comment by SharkWipf** [comment]
@k06a that's a horribly insecure idea, as any generated address comes with the private key and if you have the private key you have full access to the wallet, even after you'd sell the address.  
Anyone who'd offer such a service would most likely be scamming their users, or at the very least would be accused of it.  
You can see my Mkeys/s in my earlier post, although I haven't used this program anymore since then because of, well, this issue and the issue where it locks up after running for a while.

**Comment by k06a** [comment]
@SharkWipf my service is using split-key address generation scheme. And no one but buyer knows the complete Private Key.

https://en.bitcoin.it/wiki/Split-key_vanity_address

**Comment by SharkWipf** [comment]
@k06a Ah, didn't know that existed, interesting.  
Anyway, I've since taken apart the mining rig as it doesn't cover its own electricity bills anymore, so it's not relevant for me anymore anyway.

**Comment by adminnu** [comment]
Yes, everything is clear, how much is not saturated. Just your cnp can not cope with such a number of video cards. Run one video card, it will be 100%, and if you start to increase the number of cards, the load will fall. Again, this assumption, which I have manifested. I have 6 cards and a processor, a weak Pentium, is already on one card.

**Comment by SharkWipf** [comment]
My rig runs dual 16-thread Xeons with each 4 GPUs having a direct
connection to one CPU and each other, so I doubt it's a CPU or bandwidth
bottleneck.

**Comment by adminnu** [comment]
Then yes. This means that some parameter or initial setting does not allow sending more than a certain amount to a point on the map.

**Comment by lukechilds** [comment]
Just chiming in to say I'm also getting the same issue and it's pretty bad. Some stats to back it up:

```
1 process of 1 k80 40 Mkey/s
1 process of 8 K80s 230 Mkey/s
8 processes of 1 K80 320 Mkey/s
```

As you can see, running 8 separate processes with a single GPU gives me 100% expected throughput (8 * 40 Mkey/s = 320 Mkey/s). Running a single process with all 8 GPUs only gives me about 70% expected throughput.

**Comment by ilovelinux** [comment]
What about CPU loading? Could you tell us what's your CPU loading percentage with 1 process/1 gpu, 1 process/8 gpu and 8 process/ 1 gpu please?

**Comment by lukechilds** [comment]
Sorry, I no longer have access to the machine I was testing on. I will report back here if that changes.

**Comment by Immrspy** [comment]
Could this be related to #205 The crashing and lack of multi-gpu scaling?

**Comment by fennectech** [comment]
How the fuck can i fix this :/.  I have a machine that can reproduce this issue.
